// @yarnpkg/plugin-essentials/lib/commands/plugin/import/sources.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
const tslib_1 = require("tslib");
const cli_1 = require("@yarnpkg/cli");
const core_1 = require("@yarnpkg/core");
const core_2 = require("@yarnpkg/core");
const fslib_1 = require("@yarnpkg/fslib");
const clipanion_1 = require("clipanion");
const os_1 = require("os");
const sources_1 = require("../../set/version/sources");
const import_1 = require("../import");
const list_1 = require("../list");
const buildWorkflow = ({ pluginName, noMinify }, target) => [
    [`yarn`, `build:${pluginName}`, ...noMinify ? [`--no-minify`] : [], `|`],
];
// eslint-disable-next-line arca/no-default-export
class PluginDlSourcesCommand extends cli_1.BaseCommand {
    constructor() {
        super(...arguments);
        this.repository = `https://github.com/yarnpkg/berry.git`;
        this.branch = `master`;
        this.noMinify = false;
        this.force = false;
    }
    async execute() {
        const configuration = await core_2.Configuration.find(this.context.cwd, this.context.plugins);
        const target = typeof this.installPath !== `undefined`
            ? fslib_1.ppath.resolve(this.context.cwd, fslib_1.npath.toPortablePath(this.installPath))
            : fslib_1.ppath.resolve(fslib_1.npath.toPortablePath(os_1.tmpdir()), `yarnpkg-sources`, core_1.hashUtils.makeHash(this.repository).slice(0, 6));
        const report = await core_2.StreamReport.start({
            configuration,
            stdout: this.context.stdout,
        }, async (report) => {
            const { project } = await core_2.Project.find(configuration, this.context.cwd);
            const ident = core_1.structUtils.parseIdent(this.name.replace(/^((@yarnpkg\/)?plugin-)?/, `@yarnpkg/plugin-`));
            const identStr = core_1.structUtils.stringifyIdent(ident);
            const data = await list_1.getAvailablePlugins(configuration);
            if (!Object.prototype.hasOwnProperty.call(data, identStr))
                throw new core_2.ReportError(core_2.MessageName.PLUGIN_NAME_NOT_FOUND, `Couldn't find a plugin named "${identStr}" on the remote registry. Note that only the plugins referenced on our website (https://github.com/yarnpkg/berry/blob/master/plugins.yml) can be built and imported from sources.`);
            const pluginSpec = identStr;
            const pluginName = pluginSpec.replace(/@yarnpkg\//, ``);
            await sources_1.prepareRepo(this, { configuration, report, target });
            report.reportSeparator();
            report.reportInfo(core_2.MessageName.UNNAMED, `Building a fresh ${pluginName}`);
            report.reportSeparator();
            await sources_1.runWorkflow(buildWorkflow({
                pluginName,
                noMinify: this.noMinify,
            }, target), { configuration, context: this.context, target });
            report.reportSeparator();
            const pluginPath = fslib_1.ppath.resolve(target, `packages/${pluginName}/bundles/${pluginSpec}.js`);
            const pluginBuffer = await fslib_1.xfs.readFilePromise(pluginPath);
            await import_1.savePlugin(pluginSpec, pluginBuffer, { project, report });
        });
        return report.exitCode();
    }
}
PluginDlSourcesCommand.usage = clipanion_1.Command.Usage({
    category: `Plugin-related commands`,
    description: `build a plugin from sources`,
    details: `
      This command clones the Yarn repository into a temporary folder, builds the specified contrib plugin and updates the configuration to reference it in further CLI invocations.

      The plugins can be referenced by their short name if sourced from the official Yarn repository.
    `,
    examples: [[
            `Build and activate the "@yarnpkg/plugin-exec" plugin`,
            `$0 plugin import from sources @yarnpkg/plugin-exec`,
        ], [
            `Build and activate the "@yarnpkg/plugin-exec" plugin (shorthand)`,
            `$0 plugin import from sources exec`,
        ]],
});
tslib_1.__decorate([
    clipanion_1.Command.String()
], PluginDlSourcesCommand.prototype, "name", void 0);
tslib_1.__decorate([
    clipanion_1.Command.String(`--path`, { description: `The path where the repository should be cloned to` })
], PluginDlSourcesCommand.prototype, "installPath", void 0);
tslib_1.__decorate([
    clipanion_1.Command.String(`--repository`, { description: `The repository that should be cloned` })
], PluginDlSourcesCommand.prototype, "repository", void 0);
tslib_1.__decorate([
    clipanion_1.Command.String(`--branch`, { description: `The branch of the repository that should be cloned` })
], PluginDlSourcesCommand.prototype, "branch", void 0);
tslib_1.__decorate([
    clipanion_1.Command.Boolean(`--no-minify`, { description: `Build a plugin for development (debugging) - non-minified and non-mangled` })
], PluginDlSourcesCommand.prototype, "noMinify", void 0);
tslib_1.__decorate([
    clipanion_1.Command.Boolean(`-f,--force`, { description: `Always clone the repository instead of trying to fetch the latest commits` })
], PluginDlSourcesCommand.prototype, "force", void 0);
tslib_1.__decorate([
    clipanion_1.Command.Path(`plugin`, `import`, `from`, `sources`)
], PluginDlSourcesCommand.prototype, "execute", null);
exports.default = PluginDlSourcesCommand;
