// @yarnpkg/plugin-git/lib/GitResolver.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.GitResolver = void 0;
const tslib_1 = require("tslib");
const core_1 = require("@yarnpkg/core");
const core_2 = require("@yarnpkg/core");
const core_3 = require("@yarnpkg/core");
const gitUtils = tslib_1.__importStar(require("./gitUtils"));
class GitResolver {
    supportsDescriptor(descriptor, opts) {
        return gitUtils.isGitUrl(descriptor.range);
    }
    supportsLocator(locator, opts) {
        return gitUtils.isGitUrl(locator.reference);
    }
    shouldPersistResolution(locator, opts) {
        return true;
    }
    bindDescriptor(descriptor, fromLocator, opts) {
        return descriptor;
    }
    getResolutionDependencies(descriptor, opts) {
        return [];
    }
    async getCandidates(descriptor, dependencies, opts) {
        const reference = await gitUtils.resolveUrl(descriptor.range, opts.project.configuration);
        const locator = core_1.structUtils.makeLocator(descriptor, reference);
        return [locator];
    }
    async getSatisfying(descriptor, references, opts) {
        return null;
    }
    async resolve(locator, opts) {
        if (!opts.fetchOptions)
            throw new Error(`Assertion failed: This resolver cannot be used unless a fetcher is configured`);
        const packageFetch = await opts.fetchOptions.fetcher.fetch(locator, opts.fetchOptions);
        const manifest = await core_1.miscUtils.releaseAfterUseAsync(async () => {
            return await core_3.Manifest.find(packageFetch.prefixPath, { baseFs: packageFetch.packageFs });
        }, packageFetch.releaseFs);
        return {
            ...locator,
            version: manifest.version || `0.0.0`,
            languageName: opts.project.configuration.get(`defaultLanguageName`),
            linkType: core_2.LinkType.HARD,
            dependencies: manifest.dependencies,
            peerDependencies: manifest.peerDependencies,
            dependenciesMeta: manifest.dependenciesMeta,
            peerDependenciesMeta: manifest.peerDependenciesMeta,
            bin: manifest.bin,
        };
    }
}
exports.GitResolver = GitResolver;
