// @yarnpkg/plugin-npm-cli/lib/commands/npm/audit.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
const tslib_1 = require("tslib");
const cli_1 = require("@yarnpkg/cli");
const core_1 = require("@yarnpkg/core");
const plugin_npm_1 = require("@yarnpkg/plugin-npm");
const clipanion_1 = require("clipanion");
const npmAuditTypes = tslib_1.__importStar(require("../../npmAuditTypes"));
const npmAuditUtils = tslib_1.__importStar(require("../../npmAuditUtils"));
// eslint-disable-next-line arca/no-default-export
class AuditCommand extends cli_1.BaseCommand {
    constructor() {
        super(...arguments);
        this.all = false;
        this.recursive = false;
        this.environment = npmAuditTypes.Environment.All;
        this.json = false;
        this.severity = npmAuditTypes.Severity.Info;
    }
    async execute() {
        const configuration = await core_1.Configuration.find(this.context.cwd, this.context.plugins);
        const { project, workspace } = await core_1.Project.find(configuration, this.context.cwd);
        if (!workspace)
            throw new cli_1.WorkspaceRequiredError(project.cwd, this.context.cwd);
        await project.restoreInstallState();
        const requires = npmAuditUtils.getRequires(project, workspace, { all: this.all, environment: this.environment });
        const dependencies = npmAuditUtils.getDependencies(project, workspace, { all: this.all });
        if (!this.recursive) {
            for (const key of Object.keys(dependencies)) {
                if (!Object.prototype.hasOwnProperty.call(requires, key)) {
                    delete dependencies[key];
                }
                else {
                    dependencies[key].requires = {};
                }
            }
        }
        const body = {
            requires,
            dependencies,
        };
        const registry = plugin_npm_1.npmConfigUtils.getPublishRegistry(workspace.manifest, {
            configuration,
        });
        let result;
        const httpReport = await core_1.LightReport.start({
            configuration,
            stdout: this.context.stdout,
        }, async () => {
            try {
                result = (await plugin_npm_1.npmHttpUtils.post(`/-/npm/v1/security/audits/quick`, body, {
                    authType: plugin_npm_1.npmHttpUtils.AuthType.NO_AUTH,
                    configuration,
                    jsonResponse: true,
                    registry,
                }));
            }
            catch (err) {
                if (err.name !== `HTTPError`) {
                    throw err;
                }
                else {
                    throw new core_1.ReportError(core_1.MessageName.EXCEPTION, err.toString());
                }
            }
        });
        if (httpReport.hasErrors())
            return httpReport.exitCode();
        const hasError = npmAuditUtils.isError(result.metadata.vulnerabilities, this.severity);
        if (!this.json && hasError) {
            core_1.treeUtils.emitTree(npmAuditUtils.getReportTree(result, this.severity), {
                configuration,
                json: this.json,
                stdout: this.context.stdout,
                separators: 2,
            });
            return 1;
        }
        const outReport = await core_1.StreamReport.start({
            configuration,
            includeFooter: false,
            json: this.json,
            stdout: this.context.stdout,
        }, async (report) => {
            report.reportJson(result);
            if (!hasError) {
                report.reportInfo(core_1.MessageName.EXCEPTION, `No audit suggestions`);
            }
        });
        return outReport.exitCode();
    }
}
AuditCommand.usage = clipanion_1.Command.Usage({
    description: `perform a vulnerability audit against the installed packages`,
    details: `
      This command checks for known security reports on the packages you use. The reports are by default extracted from the npm registry, and may or may not be relevant to your actual program (not all vulnerabilities affect all code paths).

      For consistency with our other commands the default is to only check the direct dependencies for the active workspace. To extend this search to all workspaces, use \`-A,--all\`. To extend this search to both direct and transitive dependencies, use \`-R,--recursive\`.

      Applying the \`--severity\` flag will limit the audit table to vulnerabilities of the corresponding severity and above. Valid values are ${npmAuditUtils.allSeverities.map(value => `\`${value}\``).join(`, `)}.

      If the \`--json\` flag is set, Yarn will print the output exactly as received from the registry. Regardless of this flag, the process will exit with a non-zero exit code if a report is found for the selected packages.

      To understand the dependency tree requiring vulnerable packages, check the raw report with the \`--json\` flag or use \`yarn why <package>\` to get more information as to who depends on them.
    `,
    examples: [[
            `Checks for known security issues with the installed packages. The output is a list of known issues.`,
            `yarn npm audit`,
        ], [
            `Audit dependencies in all workspaces`,
            `yarn npm audit --all`,
        ], [
            `Limit auditing to \`dependencies\` (excludes \`devDependencies\`)`,
            `yarn npm audit --environment production`,
        ], [
            `Show audit report as valid JSON`,
            `yarn npm audit --json`,
        ], [
            `Audit all direct and transitive dependencies`,
            `yarn npm audit --recursive`,
        ], [
            `Output moderate (or more severe) vulnerabilities`,
            `yarn npm audit --severity moderate`,
        ]],
});
tslib_1.__decorate([
    clipanion_1.Command.Boolean(`-A,--all`)
], AuditCommand.prototype, "all", void 0);
tslib_1.__decorate([
    clipanion_1.Command.Boolean(`-R,--recursive`)
], AuditCommand.prototype, "recursive", void 0);
tslib_1.__decorate([
    clipanion_1.Command.String(`--environment`)
], AuditCommand.prototype, "environment", void 0);
tslib_1.__decorate([
    clipanion_1.Command.Boolean(`--json`)
], AuditCommand.prototype, "json", void 0);
tslib_1.__decorate([
    clipanion_1.Command.String(`--severity`)
], AuditCommand.prototype, "severity", void 0);
tslib_1.__decorate([
    clipanion_1.Command.Path(`npm`, `audit`)
], AuditCommand.prototype, "execute", null);
exports.default = AuditCommand;
