// @yarnpkg/plugin-essentials/lib/commands/exec.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
const tslib_1 = require("tslib");
const cli_1 = require("@yarnpkg/cli");
const core_1 = require("@yarnpkg/core");
const core_2 = require("@yarnpkg/core");
const fslib_1 = require("@yarnpkg/fslib");
const clipanion_1 = require("clipanion");
// eslint-disable-next-line arca/no-default-export
class ExecCommand extends cli_1.BaseCommand {
    constructor() {
        super(...arguments);
        this.args = [];
    }
    async execute() {
        const configuration = await core_1.Configuration.find(this.context.cwd, this.context.plugins);
        const { project } = await core_1.Project.find(configuration, this.context.cwd);
        return await fslib_1.xfs.mktempPromise(async (binFolder) => {
            const { code } = await core_2.execUtils.pipevp(this.commandName, this.args, {
                cwd: this.context.cwd,
                stdin: this.context.stdin,
                stdout: this.context.stdout,
                stderr: this.context.stderr,
                env: await core_2.scriptUtils.makeScriptEnv({ project, binFolder }),
            });
            return code;
        });
    }
}
ExecCommand.usage = clipanion_1.Command.Usage({
    description: `execute a shell command`,
    details: `
      This command simply executes a shell binary within the context of the root directory of the active workspace.

      It also makes sure to call it in a way that's compatible with the current project (for example, on PnP projects the environment will be setup in such a way that PnP will be correctly injected into the environment).
    `,
    examples: [[
            `Execute a shell command`,
            `$0 exec echo Hello World`,
        ]],
});
tslib_1.__decorate([
    clipanion_1.Command.String()
], ExecCommand.prototype, "commandName", void 0);
tslib_1.__decorate([
    clipanion_1.Command.Proxy()
], ExecCommand.prototype, "args", void 0);
tslib_1.__decorate([
    clipanion_1.Command.Path(`exec`)
], ExecCommand.prototype, "execute", null);
exports.default = ExecCommand;
