// @yarnpkg/plugin-node-modules/lib/index.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.getPnpPath = void 0;
const core_1 = require("@yarnpkg/core");
const fslib_1 = require("@yarnpkg/fslib");
const pnpify_1 = require("@yarnpkg/pnpify");
const NodeModulesLinker_1 = require("./NodeModulesLinker");
const PnpLooseLinker_1 = require("./PnpLooseLinker");
const getPnpPath = (project) => fslib_1.ppath.join(project.cwd, `.pnp.js`);
exports.getPnpPath = getPnpPath;
const plugin = {
    configuration: {
        nmHoistingLimits: {
            description: `Prevent packages can be hoisted past specific levels`,
            type: core_1.SettingsType.STRING,
            values: [
                pnpify_1.NodeModulesHoistingLimits.WORKSPACES,
                pnpify_1.NodeModulesHoistingLimits.DEPENDENCIES,
                pnpify_1.NodeModulesHoistingLimits.NONE,
            ],
            default: `none`,
        },
    },
    linkers: [
        NodeModulesLinker_1.NodeModulesLinker,
        PnpLooseLinker_1.PnpLooseLinker,
    ],
};
// eslint-disable-next-line arca/no-default-export
exports.default = plugin;
