// @yarnpkg/pnpify/lib/index.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.getArchivePath = exports.buildLocatorMap = exports.buildNodeModulesTree = exports.NodeModulesFS = exports.NodeModulesHoistingLimits = exports.patchFs = void 0;
const tslib_1 = require("tslib");
const fslib_1 = require("@yarnpkg/fslib");
const fs_1 = tslib_1.__importDefault(require("fs"));
const NodeModulesFS_1 = require("./NodeModulesFS");
Object.defineProperty(exports, "NodeModulesFS", { enumerable: true, get: function () { return NodeModulesFS_1.NodeModulesFS; } });
const buildNodeModulesTree_1 = require("./buildNodeModulesTree");
Object.defineProperty(exports, "getArchivePath", { enumerable: true, get: function () { return buildNodeModulesTree_1.getArchivePath; } });
const buildNodeModulesTree_2 = require("./buildNodeModulesTree");
Object.defineProperty(exports, "buildNodeModulesTree", { enumerable: true, get: function () { return buildNodeModulesTree_2.buildNodeModulesTree; } });
Object.defineProperty(exports, "buildLocatorMap", { enumerable: true, get: function () { return buildNodeModulesTree_2.buildLocatorMap; } });
const dynamicRequire_1 = require("./dynamicRequire");
let fsPatched = false;
let pnp;
try {
    pnp = dynamicRequire_1.dynamicRequire(`pnpapi`);
}
catch (e) {
}
const patchFs = () => {
    if (pnp && !fsPatched) {
        const realFs = { ...fs_1.default };
        const nodeModulesFS = new NodeModulesFS_1.NodeModulesFS(pnp, { realFs });
        fslib_1.patchFs(fs_1.default, nodeModulesFS);
        fsPatched = true;
    }
};
exports.patchFs = patchFs;
if (!process.mainModule)
    exports.patchFs();
var buildNodeModulesTree_3 = require("./buildNodeModulesTree");
Object.defineProperty(exports, "NodeModulesHoistingLimits", { enumerable: true, get: function () { return buildNodeModulesTree_3.NodeModulesHoistingLimits; } });
