// @yarnpkg/plugin-http/lib/TarballHttpResolver.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.TarballHttpResolver = void 0;
const core_1 = require("@yarnpkg/core");
const core_2 = require("@yarnpkg/core");
const core_3 = require("@yarnpkg/core");
const constants_1 = require("./constants");
class TarballHttpResolver {
    supportsDescriptor(descriptor, opts) {
        if (!constants_1.TARBALL_REGEXP.test(descriptor.range))
            return false;
        if (constants_1.PROTOCOL_REGEXP.test(descriptor.range))
            return true;
        return false;
    }
    supportsLocator(locator, opts) {
        if (!constants_1.TARBALL_REGEXP.test(locator.reference))
            return false;
        if (constants_1.PROTOCOL_REGEXP.test(locator.reference))
            return true;
        return false;
    }
    shouldPersistResolution(locator, opts) {
        return true;
    }
    bindDescriptor(descriptor, fromLocator, opts) {
        return descriptor;
    }
    getResolutionDependencies(descriptor, opts) {
        return [];
    }
    async getCandidates(descriptor, dependencies, opts) {
        return [core_3.structUtils.convertDescriptorToLocator(descriptor)];
    }
    async getSatisfying(descriptor, references, opts) {
        return null;
    }
    async resolve(locator, opts) {
        if (!opts.fetchOptions)
            throw new Error(`Assertion failed: This resolver cannot be used unless a fetcher is configured`);
        const packageFetch = await opts.fetchOptions.fetcher.fetch(locator, opts.fetchOptions);
        const manifest = await core_3.miscUtils.releaseAfterUseAsync(async () => {
            return await core_1.Manifest.find(packageFetch.prefixPath, { baseFs: packageFetch.packageFs });
        }, packageFetch.releaseFs);
        return {
            ...locator,
            version: manifest.version || `0.0.0`,
            languageName: opts.project.configuration.get(`defaultLanguageName`),
            linkType: core_2.LinkType.HARD,
            dependencies: manifest.dependencies,
            peerDependencies: manifest.peerDependencies,
            dependenciesMeta: manifest.dependenciesMeta,
            peerDependenciesMeta: manifest.peerDependenciesMeta,
            bin: manifest.bin,
        };
    }
}
exports.TarballHttpResolver = TarballHttpResolver;
