// @yarnpkg/plugin-file/lib/FileResolver.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.FileResolver = void 0;
const tslib_1 = require("tslib");
const core_1 = require("@yarnpkg/core");
const core_2 = require("@yarnpkg/core");
const core_3 = require("@yarnpkg/core");
const constants_1 = require("./constants");
const fileUtils = tslib_1.__importStar(require("./fileUtils"));
// We use this for the folders to be regenerated without bumping the whole cache
const CACHE_VERSION = 1;
class FileResolver {
    supportsDescriptor(descriptor, opts) {
        if (descriptor.range.match(constants_1.FILE_REGEXP))
            return true;
        if (!descriptor.range.startsWith(constants_1.PROTOCOL))
            return false;
        return true;
    }
    supportsLocator(locator, opts) {
        if (!locator.reference.startsWith(constants_1.PROTOCOL))
            return false;
        return true;
    }
    shouldPersistResolution(locator, opts) {
        return false;
    }
    bindDescriptor(descriptor, fromLocator, opts) {
        if (constants_1.FILE_REGEXP.test(descriptor.range))
            descriptor = core_1.structUtils.makeDescriptor(descriptor, `${constants_1.PROTOCOL}${descriptor.range}`);
        return core_1.structUtils.bindDescriptor(descriptor, {
            locator: core_1.structUtils.stringifyLocator(fromLocator),
        });
    }
    getResolutionDependencies(descriptor, opts) {
        return [];
    }
    async getCandidates(descriptor, dependencies, opts) {
        if (!opts.fetchOptions)
            throw new Error(`Assertion failed: This resolver cannot be used unless a fetcher is configured`);
        const { path, parentLocator } = fileUtils.parseSpec(descriptor.range);
        if (parentLocator === null)
            throw new Error(`Assertion failed: The descriptor should have been bound`);
        const archiveBuffer = await fileUtils.makeBufferFromLocator(core_1.structUtils.makeLocator(descriptor, core_1.structUtils.makeRange({
            protocol: constants_1.PROTOCOL,
            source: path,
            selector: path,
            params: {
                locator: core_1.structUtils.stringifyLocator(parentLocator),
            },
        })), { protocol: constants_1.PROTOCOL, fetchOptions: opts.fetchOptions });
        const folderHash = core_1.hashUtils.makeHash(`${CACHE_VERSION}`, archiveBuffer).slice(0, 6);
        return [fileUtils.makeLocator(descriptor, { parentLocator, path, folderHash, protocol: constants_1.PROTOCOL })];
    }
    async getSatisfying(descriptor, references, opts) {
        return null;
    }
    async resolve(locator, opts) {
        if (!opts.fetchOptions)
            throw new Error(`Assertion failed: This resolver cannot be used unless a fetcher is configured`);
        const packageFetch = await opts.fetchOptions.fetcher.fetch(locator, opts.fetchOptions);
        const manifest = await core_1.miscUtils.releaseAfterUseAsync(async () => {
            return await core_3.Manifest.find(packageFetch.prefixPath, { baseFs: packageFetch.packageFs });
        }, packageFetch.releaseFs);
        return {
            ...locator,
            version: manifest.version || `0.0.0`,
            languageName: opts.project.configuration.get(`defaultLanguageName`),
            linkType: core_2.LinkType.HARD,
            dependencies: manifest.dependencies,
            peerDependencies: manifest.peerDependencies,
            dependenciesMeta: manifest.dependenciesMeta,
            peerDependenciesMeta: manifest.peerDependenciesMeta,
            bin: manifest.bin,
        };
    }
}
exports.FileResolver = FileResolver;
