// @yarnpkg/plugin-file/lib/TarballFileResolver.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.TarballFileResolver = void 0;
const core_1 = require("@yarnpkg/core");
const core_2 = require("@yarnpkg/core");
const core_3 = require("@yarnpkg/core");
const fslib_1 = require("@yarnpkg/fslib");
const constants_1 = require("./constants");
class TarballFileResolver {
    supportsDescriptor(descriptor, opts) {
        if (!constants_1.TARBALL_REGEXP.test(descriptor.range))
            return false;
        if (descriptor.range.startsWith(constants_1.PROTOCOL))
            return true;
        if (constants_1.FILE_REGEXP.test(descriptor.range))
            return true;
        return false;
    }
    supportsLocator(locator, opts) {
        if (!constants_1.TARBALL_REGEXP.test(locator.reference))
            return false;
        if (locator.reference.startsWith(constants_1.PROTOCOL))
            return true;
        return false;
    }
    shouldPersistResolution(locator, opts) {
        return true;
    }
    bindDescriptor(descriptor, fromLocator, opts) {
        if (constants_1.FILE_REGEXP.test(descriptor.range))
            descriptor = core_3.structUtils.makeDescriptor(descriptor, `${constants_1.PROTOCOL}${descriptor.range}`);
        return core_3.structUtils.bindDescriptor(descriptor, {
            locator: core_3.structUtils.stringifyLocator(fromLocator),
        });
    }
    getResolutionDependencies(descriptor, opts) {
        return [];
    }
    async getCandidates(descriptor, dependencies, opts) {
        let path = descriptor.range;
        if (path.startsWith(constants_1.PROTOCOL))
            path = path.slice(constants_1.PROTOCOL.length);
        return [core_3.structUtils.makeLocator(descriptor, `${constants_1.PROTOCOL}${fslib_1.npath.toPortablePath(path)}`)];
    }
    async getSatisfying(descriptor, references, opts) {
        return null;
    }
    async resolve(locator, opts) {
        if (!opts.fetchOptions)
            throw new Error(`Assertion failed: This resolver cannot be used unless a fetcher is configured`);
        const packageFetch = await opts.fetchOptions.fetcher.fetch(locator, opts.fetchOptions);
        const manifest = await core_3.miscUtils.releaseAfterUseAsync(async () => {
            return await core_1.Manifest.find(packageFetch.prefixPath, { baseFs: packageFetch.packageFs });
        }, packageFetch.releaseFs);
        return {
            ...locator,
            version: manifest.version || `0.0.0`,
            languageName: opts.project.configuration.get(`defaultLanguageName`),
            linkType: core_2.LinkType.HARD,
            dependencies: manifest.dependencies,
            peerDependencies: manifest.peerDependencies,
            dependenciesMeta: manifest.dependenciesMeta,
            peerDependenciesMeta: manifest.peerDependenciesMeta,
            bin: manifest.bin,
        };
    }
}
exports.TarballFileResolver = TarballFileResolver;
